### PR TITLE
Add Gas Price to /v1/chains/<chain_id>

### LINKED ISSUE #444: Add Gas Price Oracle to ChainInfo
The Safe Config Service now returns the chain gas price oracle for each chain:

```javascript
{
  "gasPriceOracle": {
    "url": <string>,
    "gasParameter": <string>
  }
}
```

The gas price oracle should be exposed to the clients via GET /v1/chains/<chain_id>

### PR DESCRIPTION
**Pull request by fmrsabino**
Closes #444 

- Add Gas Price to `ChainInfo`

### CODE REVIEW

File: src/models/chains.rs [hunk 1]
```diff
@@ -18,3 +20,17 @@ pub struct NativeCurrency {
...
+pub struct Theme {
+    pub text_color: String,
+    pub background_color: String,
+}
+
+#[derive(Serialize, Deserialize, Debug, PartialEq, Clone)]
+#[serde(rename_all = "camelCase")]
+pub struct GasPriceOracle {
+    pub url: String,
+    pub gas_parameter: Option<String>,
```
Comment (rmeissner) [inline src/models/chains.rs:35]: Is this the only format we will have? 

Afaik the web version also supports having a constant value set. And there was the discussion about a multiplier for the oracle solution.

**Review by rmeissner — CHANGES_REQUESTED** [review]

**Comment by fmrsabino** [comment]
Blocked by: https://github.com/gnosis/safe-config-service/issues/87#issuecomment-872203630

File: src/models/chains.rs (hunk 1 above)
Comment (jpalvarezl) [reply src/models/chains.rs:35]: If that is the situation we should probably change this from a `struct` into a `enum` and make it polymorphic.

Comment (fmrsabino) [reply src/models/chains.rs:35]: The `gas_parameter` can have any `String` value (it'd depend on the oracle itself)

**Comment by fmrsabino** [comment]
Config Service PR: https://github.com/gnosis/safe-config-service/pull/146

File: src/models/chains.rs [hunk 2]
```diff
@@ -27,3 +28,11 @@ pub struct Theme {
     pub text_color: String,
     pub background_color: String,
 }
+
+#[derive(Serialize, Deserialize, Debug, PartialEq, Clone)]
```
Comment (jpalvarezl) [inline src/models/chains.rs:32]: We probably don't need `Clone`, right?

**Review by jpalvarezl — APPROVED** [review]
LGTM , I might get rid of `Clone` derive if it's not needed, but not a mandatory change :)

File: src/models/chains.rs [hunk 3]
```diff
@@ -28,3 +29,11 @@ pub struct Theme {
     pub text_color: String,
     pub background_color: String,
 }
+
+#[derive(Serialize, Deserialize, Debug, PartialEq)]
+#[serde(rename_all = "camelCase")]
+pub struct GasPriceOracle {
+    pub url: String,
+    pub gas_parameter: Option<String>,
+    pub gwei_factor: String,
```
Comment (jpalvarezl) [inline src/models/chains.rs:38]: The `gwei_factor` is always present or does it have any default value ?

Comment (fmrsabino) [reply src/models/chains.rs:38]: Always present. Default value from the config service is `"1.000000000"`

**Review by jpalvarezl — APPROVED** [review]
LGTM, although I am not entirely aware of the web team expectation. If that's all covered then I would say we are good to merge.

**Comment by fmrsabino** [comment]
@rmeissner I need your review 😄 🙏

File: src/models/service/chains.rs [hunk 4]
```diff
@@ -28,3 +30,17 @@ pub struct Theme {
     pub text_color: String,
     pub background_color: String,
 }
+
+#[derive(Serialize, Debug, PartialEq, Clone)]
+#[serde(tag = "type")]
+#[serde(rename_all = "lowercase")]
```
Comment (jpalvarezl) [inline src/models/service/chains.rs:36]: Clients generally expect enum variants as `SCREAMING_SNAKE_CASE`

File: src/models/backend/chains.rs [hunk 5]
```diff
@@ -28,3 +29,17 @@ pub struct Theme {
...
+#[serde(rename_all = "lowercase")]
+pub enum GasPrice {
+    #[serde(rename_all = "camelCase")]
+    Oracle {
+        url: String,
+        gas_parameter: Option<String>,
+        gwei_factor: String,
+    },
+    #[serde(rename_all = "camelCase")]
+    Fixed { wei_value: String },
```
Comment (jpalvarezl) [inline src/models/backend/chains.rs:44]:
```suggestion
    Fixed { wei_value: String },
    #[serde(other)]
    Unknown,
```

Just to protect ourselves from ourselves in the future :D

**Review by jpalvarezl — APPROVED** [review]
I would approve as to not block you, but I think it would be good to implement my suggestions. `serde tag` will try to match the underlying structure of the variant, so testing for `unknown` would be nice too. Just by adding some wacky structure.

File: src/models/service/chains.rs (hunk 4 above)
Comment (fmrsabino) [reply src/models/service/chains.rs:36]: Good catch. The config service exposes this as a lower case but this can be different when serialising

**Comment by fmrsabino** [comment]
> I would approve as to not block you, but I think it would be good to implement my suggestions. `serde tag` will try to match the underlying structure of the variant, so testing for `unknown` would be nice too. Just by adding some wacky structure.

Good point. Will add a test for this 👍

**Review by jpalvarezl — APPROVED** [review]

File: src/models/backend/chains.rs [hunk 6]
```diff
@@ -28,3 +29,19 @@ pub struct Theme {
...
 }
+
+#[derive(Deserialize, Debug, PartialEq, Clone)]
+#[serde(tag = "type")]
+#[serde(rename_all = "lowercase")]
+pub enum GasPrice {
+    #[serde(rename_all = "camelCase")]
+    Oracle {
+        url: String,
+        gas_parameter: Option<String>,
```
Comment (rmeissner) [inline src/models/backend/chains.rs:40]: Why is this optional? I was thinking about if there is a default used by the clients :thinking: or does `null` imply something else on config service side?

In the typescript definition it is not nullable: https://github.com/gnosis/safe-react/blob/dev/src/config/networks/network.d.ts#L74

Comment (fmrsabino) [reply src/models/backend/chains.rs:40]: It's optional (safe-config side) https://github.com/gnosis/safe-config-service/issues/87#issuecomment-865994250. I can change that.

Comment (fmrsabino) [reply src/models/backend/chains.rs:40]: Since this matches what the safe-config service returns I'd say to tackle this in a different PR

Comment (fmrsabino) [reply src/models/backend/chains.rs]: Gas parameter is now required: 86fb7a5

**Comment by fmrsabino** [comment]
Conflicts fixed cc/ @jpalvarezl @rmeissner

**Review by rmeissner — APPROVED** [review]

**Pull request merged**